modules/oauth-api/src/main/java/org/mule/runtime/oauth/api/builder/OAuthAuthorizationCodeDancerBuilder.java

```diff
@@ -115,4 +120,26 @@
    */
   OAuthAuthorizationCodeDancerBuilder externalCallbackUrl(String externalCallbackUrl);
 
+  /**
+   * Allows custom code to be run just before doing the request to the provided {@code tokenUrl}.
+   * <p>
+   * The map returned by the provided function will be passed to the {@link #afterDanceCallback(BiConsumer)}, if set.
+   * 
+   * @param callback a {@link Function} that receives the parameters that will be used in the executing dance and returns a map to
+   *        be then passed to the {@link #afterDanceCallback(BiConsumer)}. Not null.
+   * 
+   * @return this builder
+   */
+  OAuthAuthorizationCodeDancerBuilder beforeDanceCallback(Function<AuthorizationCodeRequest, Map<String, TypedValue>> callback);
```

Comment:
this is very weird. Why would I return a map that is passed to the after but is ignored by all other components? Doesn't this overlap with the state parameter?

What's the use case for this? Please clarify because otherwise, it seems like Function is a too generic contract for whatever is it you're trying to achieve here.